### PR TITLE
STITCH-156 Add tests for Android SDK

### PR DESCRIPTION
**Pull request by jsflax**
Added Kotlin for testing.

Added RESTMock framework to mock network calls: https://github.com/andrzejchm/RESTMock .

Testing public methods of `StitchClient` class, as well as general authentication and pipeline flows.

### CODE REVIEW

File: .idea/markdown-exported-files.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edaniels) [inline .idea/markdown-exported-files.xml]: Need this file?

File: build.gradle [hunk 2]
```diff
@@ -15,6 +17,7 @@ buildscript {
 
 allprojects {
     repositories {
+        maven { url "https://jitpack.io" }
```
Comment (edaniels) [inline build.gradle]: What's this for?

File: .idea/misc.xml [hunk 3]
```diff
@@ -37,7 +37,7 @@
     <ConfirmationsSetting value="0" id="Add" />
     <ConfirmationsSetting value="0" id="Remove" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" default="true" assert-keyword="true" jdk-15="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" assert-keyword="true" jdk-15="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Comment (edaniels) [inline .idea/misc.xml]: Should find a way to stop this from changing :(

File: sdk/build.gradle [hunk 4]
```diff
@@ -15,7 +16,8 @@ android {
         versionCode 20170505
         versionName "0.1.0"
 
-        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+        multiDexEnabled true
```
Comment (edaniels) [inline sdk/build.gradle]: Will using `multiDexEnabled` affect apps importing the SDK or no?

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 5]
```diff
@@ -0,0 +1,218 @@
+package com.mongodb.stitch
+
+import android.content.Context
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Looks like these imports need to be auto-organized

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+import io.appflate.restmock.android.AndroidAssetsFileParser
+import io.appflate.restmock.RESTMockServerStarter
+import io.appflate.restmock.utils.RequestMatchers.pathEndsWith
+
+/**
+ * Test public methods of StitchClient, running through authentication
+ * and pipeline flows.
+ */
+@RunWith(AndroidJUnit4::class)
+class StitchClientTest {
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Does this need a package access specifier or is it unnecessary?

Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Oh this is Kotlin so the answer may very well be no

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    companion object {
+        const val FakeAccessToken = "fake-access-token"
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: is it a kotlin convention to uppercase the first letter of a constant?

**Comment by edaniels** [comment]
cc: @mburdette if you want to skim it

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                        "metadataFields" to arrayOf(
+                                FakeMetadataFieldEmail,
+                                FakeMetadataFieldBirthday
+                        )
+                )
+        )).toString()
+
+        /** Mock data for auth object from [StitchClient.logInWithProvider] */
+        val mockAuthData = JSONObject(mapOf(
+                "accessToken" to FakeAccessToken,
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Should we use constant values for these field names in general?

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        )).toString()
+    }
+
+    /** Base context from test runner */
+    private val instrumentationCtx: Context = InstrumentationRegistry.getContext()
+    /** [StitchClient] for this test */
+    private var stitchClient: StitchClient? = null
+
+    @Before
+    fun setup() {
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: so fun

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        // start the mock server
+        RESTMockServerStarter.startSync(AndroidAssetsFileParser(instrumentationCtx), AndroidLogger())
+
+        // instantiate a new StitchClient using a dummy name and the mock baseUrl
+        val stitchClient = StitchClient(instrumentationCtx, "dummy-app", RESTMockServer.getUrl())
+        this.stitchClient = stitchClient
+
+        // clear all instances of the internal [SharedPreferences] to start with a clean slate
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: What's this convention of using brackets around the class name?

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        stitchClient.properties.clear()
+        stitchClient.javaClass.kotlin.declaredMemberProperties.first {
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Can you explain what's going on here?

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+            it.name == "_preferences"
+        }.also { it.isAccessible = true }.get(stitchClient).also {
+            (it as SharedPreferences).edit().clear().commit()
+        }
+
+        // mock out all calls related to auth
+        RESTMockServer.whenPOST(pathContains("auth")).thenReturnString(mockAuthData)
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: I'd like these to be more specific as spec-like such that we know exactly what type of path we are looking for. It may be easy to mock the wrong thing this way.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        RESTMockServer.whenGET(pathEndsWith("auth")).thenReturnString(mockProviderData)
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Will these types of responses return JSON content-types or what?

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            }
+        })
+
+        // assert that we have not authenticated yet
+        assertThat(!stitchClient!!.isAuthenticated)
+
+        // fetch mocked authProviders
+        await(stitchClient!!.authProviders).let {
+            // assert all providers provided in mock are available
+            assertThat(it.hasAnonymous() && it.hasEmailPassword()
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: If we have a test like this we should have another test where there is no provider available and a partial amount available to cover our edge cases

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+            when (it) {
+                is GoogleAuthProviderInfo ->
+                    assertThat(it.clientId == FakeGoogleClientId)
+                is FacebookAuthProviderInfo ->
+                    assertThat(it.applicationId == FakeFacebookClientId)
+            }
+        }
+
+        // log in anonymously
+        await(stitchClient!!.logInWithProvider(AnonymousAuthProvider()))
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: !! is similar to getting the optional and dereferencing?

File: sdk/src/androidTest/java/com/mongodb/stitch/TestUtils.kt [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+/**
+ * Convenience method that wraps a [Task] in the [latch] method, awaiting the result.
+ *
+ * @param[task] pending task to wait for
+ *
+ * @return result of the task after it's been completed
+ */
+fun <T> await(task: Task<T>): T {
```
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/TestUtils.kt]: so is using coroutines not a better idea with async and suspend?

File: .idea/misc.xml (hunk 3 above)
Comment (jsflax) [inline .idea/misc.xml]: Will look into this. Could just ignore the file.

File: build.gradle (hunk 2 above)
Comment (jsflax) [inline build.gradle]: This is for the RESTMock framework.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 6 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Unnecessary, public by default.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 7 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Looks like I was still thinking Swift-y here. Should be defaulting to Java naming conventions here: http://kotlinlang.org/docs/reference/coding-conventions.html will change.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 8 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: We could, but they would be unused at the moment, as we aren't going to be accessing the mock data by keys, but in their mapped object form.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 10 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Brackets is the equivalent of `{@link SharedPreferences}` in Java

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 11 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: I'm reflectively clearing out the `_properties` member of the StitchClient.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 13 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Unsure, but they are currently being handled correctly. If I want to return more precise responses, I have to instantiate a `MockResponse` and provide that information. Will change for precision.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 12 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Will change.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 15 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Correct, `!!` is the unsafe way to unwrap optionals. In this case, it's safe as we know 100% that the `stitchClient` will be non-null and unchanged.

File: sdk/src/androidTest/java/com/mongodb/stitch/TestUtils.kt (hunk 16 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/TestUtils.kt]: https://kotlinlang.org/docs/reference/coroutines.html So after fiddling around with co-routines, I ran into a few issues, namely that they don't play well with the Tasks/Futures that we are currently using:

```kotlin
    inline fun <R> latch(count: Int = 1, block: CountDownLatch.() -> R): R {
        val latch = CountDownLatch(count)
        val r = latch.block()
        latch.await()
        return r
    }

    private suspend fun <T> fetch(task: Task<T>): T {
        latch { task.addOnCompleteListener { this.countDown() } }

        return task.result
    }

    @Test fun testAuth() = runBlocking {
        val deferredAuth = async(CommonPool) {
            fetch(stitchClient!!.logInWithProvider(AnonymousAuthProvider()))
        }

        val auth = deferredAuth.await()

        assert(auth.accessToken == FakeAccessToken)
    }
```

That was my experimentation. I ended up being able to accomplish all that in two lines without using coroutines:

```kotlin
@Test fun testAuth() {
        val auth = fetch(stitchClient!!.logInWithProvider(AnonymousAuthProvider()))

        assertThat(auth.accessToken == FakeAccessToken)
}
```

It looks like to take advantage of coroutines, we'd need to write our own Promise class in kotlin that uses the `suspended` keyword. It's possible we can implement our own `Task` class (`Task` is abstract after all), but that seems like overkill right now.

Could be a cool thought for a side project. Kotlin promises would be nice while the language is still in its nascent stages.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 5 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Will fix.

File: sdk/build.gradle (hunk 4 above)
Comment (jsflax) [inline sdk/build.gradle]: I'm going to have to test this against a sample app before I can properly answer this question.

File: sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt (hunk 14 above)
Comment (jsflax) [inline sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt]: Should also probably make sure the `StitchClient`is catching error responses (for mock errors) while we're at it. Will add.

File: sdk/src/androidTest/java/com/mongodb/stitch/TestUtils.kt (hunk 16 above)
Comment (edaniels) [inline sdk/src/androidTest/java/com/mongodb/stitch/TestUtils.kt]: makes sense to not do it then

**Comment by jsflax** [comment]
@edaniels PTAL. I've tested this against MongoCipher after a local deploy to ensure that multidex/kotlin will not have negative effects. The only change I didn't tend to was changing that `path` mocks. Thanks to the HTTP verb-age, there will not be any conflicts... for now. But it's something we'll need to keep an eye on.
